docs/shaders.rst

```diff
@@ -0,0 +1,325 @@
+    class SomeShader(WorldObjectShader):
+
+        type = "render"  # must be "render" or "compute"
+
+        def get_resources(self, wobject, shared):
+            bindings = {
+                0: Binding("u_stdinfo", "buffer/uniform", shared.uniform_buffer),
+                ...
+            }
+            self.define_bindings(0, bindings)
+            return {
+                "index_buffer": None,
+                "vertex_buffers": {},
+                "bindings": {
+                    0: bindings,
+                },
+            }
+
+        def get_pipeline_info(self, wobject, shared):
+            return {
+                "primitive_topology": wgpu.PrimitiveTopology.triangle_list,
+                "cull_mode": wgpu.CullMode.none,
+            }
+
+        def get_render_info(self, wobject, shared):
+            n_vertices = ...
+            n_instances = 1
+            render_mask = wobject.render_mask
+            if not render_mask:
+                render_mask = RenderMask.all
+            return {
+                "indices": (n_vertices, n_instances),
+                "render_mask": render_mask,
+            }
+
+        def get_code(self):
+            return (
+                self.code_definitions()
+                + self.code_common()
+                + self.code_vertex()
+                + self.code_fragment()
+            )
+
+        def code_vertex(self):
+            return """
+            @stage(vertex)
+            fn vs_main(@builtin(vertex_index) index: u32) -> @builtin(position) vec4<f32> {
+                ...
+            }
+            """
+
+        def code_fragment(self):
+            return """
+            @stage(fragment)
+            fn fs_main() -> FragmentOutput {
+               ...
+            }
+            """
+
+
+Remarks:
+
+* In ``get_resources()``, the ``Binding`` object is used to collect all the required information on a binding.
+* The wgsl code to define a group of bindings can be easily generated using ``define_bindings()``.
+* You can also manually define the wgsl code for a binding in cases where this is easier.
+  We recommend using a separate bindgroup for that.
+* By convention, methods that return wgsl code are prefixed with "code".
+* The ``render_mask`` specifies in what passes the object must be drawn. Users
+  can set it on the object, but by default it is "auto" (zero), in which case it must
+  be set by the shader. In the code above it is set to "all" which is a safe option, but if
+  the shader knows that all fragments are opaque or all fragments are transparent,
+  the ``render_mask`` can be set accordingly.
+
+
+WGSL code and templating
+------------------------
+
+The shader code is written in `WGSL <https://www.w3.org/TR/WGSL/>`_. We use `jinja2-templating <https://jinja.palletsprojects.com/>`_
+to allow flexible code generation. Here's an example:
+
+.. code-block:: python
+
+        def get_resources(self, wobject, shared):
+            self["scale"] = 1.2
```

Comment:
is there an important reason this uses brackets instead of dots to access the attribute?

Reply:
This is the way to set templating variables. Added a comment for clarification.